lib/interface-types/src/ast.rs

```diff
@@ -0,0 +1,173 @@
+    Any,
+
+    /// A string.
+    String,
+
+    /// A sequence.
+    Seq,
+
+    /// A 32-bits integer.
+    I32,
+
+    /// A 64-bits integer.
+    I64,
+
+    /// A 32-bits float.
+    F32,
+
+    /// A 64-bits float.
+    F64,
+
+    /// An `any` reference.
+    AnyRef,
+}
+
+/// Represents the kind of adapter.
+#[derive(PartialEq, Debug)]
+pub(crate) enum AdapterKind {
+    /// An adapter defined for an imported function of a WebAssembly instance.
+    Import,
+
+    /// An adapter defined for an exported function of a WebAssembly instance.
+    Export,
+
+    /// A helper function.
+    HelperFunction,
+}
+
+/// Represents an exported function signature.
+#[derive(PartialEq, Debug)]
+pub struct Export<'input> {
+    /// The function name.
+    pub name: &'input str,
+
+    /// The function input types.
+    pub input_types: Vec<InterfaceType>,
+
+    /// The function output types.
+    pub output_types: Vec<InterfaceType>,
+}
+
+/// Represents an imported function signature.
+#[derive(PartialEq, Debug)]
+pub struct Import<'input> {
+    /// The function namespace.
+    pub namespace: &'input str,
+
+    /// The function name.
+    pub name: &'input str,
+
+    /// The function input types.
+    pub input_types: Vec<InterfaceType>,
+
+    /// The function output types.
+    pub output_types: Vec<InterfaceType>,
+}
+
+/// Represents a structural type.
+#[derive(PartialEq, Debug)]
+pub struct Type<'input> {
+    /// The type name.
+    pub name: &'input str,
+
+    /// The field names.
+    pub fields: Vec<&'input str>,
+
+    /// The field types.
+    pub types: Vec<InterfaceType>,
+}
+
+/// Represents an adapter.
+#[derive(PartialEq, Debug)]
+pub enum Adapter<'input> {
+    /// An adapter for an imported function.
+    Import {
+        /// The function namespace.
+        namespace: &'input str,
+
+        /// The function name.
+        name: &'input str,
+
+        /// The function input types.
+        input_types: Vec<InterfaceType>,
+
+        /// The function output types.
+        output_types: Vec<InterfaceType>,
+
+        /// The instructions of the adapter.
+        instructions: Vec<Instruction<'input>>,
+    },
+
+    /// An adapter for an exported function.
+    Export {
+        /// The function name.
+        name: &'input str,
+
+        /// The function input types.
+        input_types: Vec<InterfaceType>,
+
+        /// The function output types.
+        output_types: Vec<InterfaceType>,
+
+        /// The instructions of the adapter.
+        instructions: Vec<Instruction<'input>>,
+    },
+
+    /// An adapter for a helper function.
+    HelperFunction {
+        /// The helper name.
+        name: &'input str,
+
+        /// The helper input types.
+        input_types: Vec<InterfaceType>,
+
+        /// The helper output types.
+        output_types: Vec<InterfaceType>,
+
+        /// The instructions of the adapter.
+        instructions: Vec<Instruction<'input>>,
+    },
+}
+
+/// Represented a forwarded export.
```

Comment:
It's not super defined in the spec yet :-D. It should be used to “rename” an export, like `main` to `_start`. I'm not sure it will be kept in the future.